Create rule S7410: Keyboard cache should be disabled for password inputs (SONARKT-583) (#4724)
SonarSource/rspec: rules/S7410/kotlin/rule.adoc

> Mobile OS uses software keyboards that provide text predictions and suggestions. It rely on caching text inputs in a local file in order to speed up typing and recall frequent phrases. When users type sensitive data into a text field where keyboard cache is enabled, the data will be exposed in clear-text in a local file. It will keep appearing in the keyboard suggestion list until the cache is cleared.

Reviewer: ```suggestion
Mobile OSes use software keyboards that provide text predictions and suggestions. These keyboards cache text inputs in a local file in order to speed up typing and to recall frequent phrases. When users type sensitive data into a text field where keyboard cache is enabled, the data will be stored in clear-text in a local file. It will keep appearing in the keyboard suggestion list until the cache is cleared.
```
I replaced some spelling errors. Apart from that, since this is a rule about passwords, maybe we can just use that instead of sensitive data.

== Why is this an issue?

> Keyboard caches are not designed to store sensitive information. Data they contain is not encrypted and can be exposed in device backups or when the device is shared.

Reviewer: This section is a bit empty, I'd suggest to either enhance this a bit or to move some content from the summary to here.

== How to fix it in Jetpack Compose

=== Code examples

==== Noncompliant code example

[source,kotlin,diff-id=1,diff-type=noncompliant]
----
TextField(
    value = text,
    onValueChange = { text = it },
    label = { Text("Password") },
    visualTransformation = PasswordVisualTransformation(),
    keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Text) // Noncompliant: keyboard cache is enabled
)
----

==== Compliant solution

[source,kotlin,diff-id=1,diff-type=compliant]
----

> TextField(

Reviewer: Prefer `SecureTextField` as a solution for this case

Reviewer 2: As discussed: this composent is still [in Alpha](https://developer.android.com/jetpack/androidx/releases/compose-material3#1.4.0-alpha01)
We should update the code example once it's official part of the stable version
